File: commons/src/main/scala/org/dinky/udf/scala/SumFunction.scala
package org.dinky.udf.scala

import org.apache.flink.table.functions.ScalarFunction

import scala.annotation.varargs

// 有多个重载求值方法的函数
class SumFunction extends ScalarFunction {

  def eval(a: Integer, b: Integer): Integer = {
    a + b
  }

  def eval(a: String, b: String): Integer = {
    Integer.valueOf(a) + Integer.valueOf(b)
  }

  @varargs // generate var-args like Java
  def eval(d: Double*): Integer = {
    d.sum.toInt
  }
}

File: commons/src/main/java/org/dinky/udf/SumFunction.java
package org.dinky.udf;

import org.apache.flink.table.functions.ScalarFunction;

/**
 * @author ZackYoung
 * @since 1.0.0
 */
public class SumFunction extends ScalarFunction {

    public Integer eval(Integer a, Integer b) {
        return a + b;
    }

    public Integer eval(String a, String b) {
        return Integer.valueOf(a) + Integer.valueOf(b);
    }

    public Integer eval(Double... d) {
        double result = 0;
        for (double value : d) {
            result += value;
        }
        return (int) result;
    }
}

